core/templates/dev/head/services/SearchService.ts

```diff
@@ -16,206 +16,225 @@
-            itemsType = 'categories';
-          } else if (urlComponent.indexOf('language_code') === 0) {
-            itemsType = 'languageCodes';
-          } else {
-            continue;
-          }
+    if (successCallback) {
+      successCallback();
+    }
+  }
 
-          try {
-            updateSearchFields(itemsType, urlComponent, selectionDetails);
-          } catch (error) {
-            selectionDetails[itemsType].selections = {};
-          }
-        }
+  isSearchInProgress(): boolean {
+    return this.numSearchesInProgress > 0;
+  }
+  // The following takes in the url search component as an argument and the
+  // selectionDetails. It will update selectionDetails with the relevant
+  // fields that were extracted from the url. It returns the unencoded
+  // search query string.
+  updateSearchFieldsBasedOnUrlQuery(
+      urlComponent: string, selectionDetails: Object): string {
+    let urlQuery = urlComponent.substring('?q='.length);
+    // The following will split the urlQuery into 3 components:
+    // 1. query
+    // 2. categories (optional)
+    // 3. language codes (default to 'en')
+    let querySegments = urlQuery.split('&');
+
+    for (let i = 1; i < querySegments.length; i++) {
+      urlComponent = decodeURIComponent(querySegments[i]);
+
+      let itemsType = null;
+      if (urlComponent.indexOf('category') === 0) {
+        itemsType = 'categories';
+      } else if (urlComponent.indexOf('language_code') === 0) {
+        itemsType = 'languageCodes';
+      } else {
+        continue;
+      }
 
-        return decodeURIComponent(querySegments[0]);
-      },
-      getCurrentUrlQueryString: function() {
-        return this.getSearchUrlQueryString(
-          _lastQuery, _lastSelectedCategories, _lastSelectedLanguageCodes);
-      },
-      loadMoreData: function(successCallback, failureCallback) {
-        // If a new query is still being sent, or the end of the page has been
-        // reached, do not fetch more results.
-        if (_isCurrentlyFetchingResults || hasReachedEndOfPage()) {
-          failureCallback(hasReachedEndOfPage());
-          return;
-        }
+      try {
+        this.updateSearchFields(itemsType, urlComponent, selectionDetails);
+      } catch (error) {
+        selectionDetails[itemsType].selections = {};
+      }
+    }
 
-        var queryUrl = getQueryUrl(this.getCurrentUrlQueryString());
+    return decodeURIComponent(querySegments[0]);
+  }
 
-        if (_searchCursor) {
-          queryUrl += '&cursor=' + _searchCursor;
-        }
+  getCurrentUrlQueryString(): string {
+    return this.getSearchUrlQueryString(
+      this._lastQuery, this._lastSelectedCategories,
+      this._lastSelectedLanguageCodes);
+  }
 
-        _isCurrentlyFetchingResults = true;
-        $http.get(queryUrl).then(function(response) {
-          _searchCursor = response.data.search_cursor;
-          _isCurrentlyFetchingResults = false;
+  loadMoreData(successCallback, failureCallback): void {
```

Comment:
What's the type of the args?

Reply:
They are callback functions....also they are being called in the code block below..so Function

Reply:
Please go ahead and type them as such then :)

Reply:
Ok done